WalletWasabi/WabiSabi/Backend/Rounds/Arena.cs

```diff
@@ -509,34 +511,39 @@ public async Task<ReissueCredentialResponse> ReissuanceAsync(ReissueCredentialRe
 				}
+			}
 
-				if (request.RealAmountCredentialRequests.Delta != 0)
-				{
-					throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.DeltaNotZero, $"Round ({round.Id}): Amount credentials delta must be zero.");
-				}
-
-				if (request.RealAmountCredentialRequests.Requested.Count() != ProtocolConstants.CredentialNumber)
-				{
-					throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.WrongNumberOfCreds, $"Round ({round.Id}): Incorrect requested number of amount credentials.");
-				}
+			if (request.RealAmountCredentialRequests.Delta != 0)
+			{
+				throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.DeltaNotZero, $"Round ({round.Id}): Amount credentials delta must be zero.");
+			}
 
-				if (request.RealVsizeCredentialRequests.Requested.Count() != ProtocolConstants.CredentialNumber)
-				{
-					throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.WrongNumberOfCreds, $"Round ({round.Id}): Incorrect requested number of weight credentials.");
-				}
+			if (request.RealAmountCredentialRequests.Requested.Count() != ProtocolConstants.CredentialNumber)
+			{
+				throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.WrongNumberOfCreds, $"Round ({round.Id}): Incorrect requested number of amount credentials.");
+			}
 
-				var commitRealAmountCredentialResponse = round.AmountCredentialIssuer.PrepareResponse(request.RealAmountCredentialRequests);
-				var commitRealVsizeCredentialResponse = round.VsizeCredentialIssuer.PrepareResponse(request.RealVsizeCredentialRequests);
-				var commitZeroAmountCredentialResponse = round.AmountCredentialIssuer.PrepareResponse(request.ZeroAmountCredentialRequests);
-				var commitZeroVsizeCredentialResponse = round.VsizeCredentialIssuer.PrepareResponse(request.ZeroVsizeCredentialsRequests);
-
-				return new(
-					commitRealAmountCredentialResponse.Commit(),
-					commitRealVsizeCredentialResponse.Commit(),
-					commitZeroAmountCredentialResponse.Commit(),
-					commitZeroVsizeCredentialResponse.Commit()
-					);
+			if (request.RealVsizeCredentialRequests.Requested.Count() != ProtocolConstants.CredentialNumber)
+			{
+				throw new WabiSabiProtocolException(WabiSabiProtocolErrorCode.WrongNumberOfCreds, $"Round ({round.Id}): Incorrect requested number of weight credentials.");
 			}
+
+			var realAmountTask = Task.Run(() => round.AmountCredentialIssuer.PrepareResponse(request.RealAmountCredentialRequests), cancellationToken);
+			var realVsizeTask = Task.Run(() => round.VsizeCredentialIssuer.PrepareResponse(request.RealVsizeCredentialRequests), cancellationToken);
+			var zeroAmountTask = Task.Run(() => round.AmountCredentialIssuer.PrepareResponse(request.ZeroAmountCredentialRequests), cancellationToken);
+			var zeroVsizeTask = Task.Run(() => round.VsizeCredentialIssuer.PrepareResponse(request.ZeroVsizeCredentialsRequests), cancellationToken);
```

Comment:
Instead of doing it in this way I would prefer to change the signature of the `PrepareResponse` from:

```c#
public ICommitableCredentialsResponse PrepareResponse(CredentialsRequest registrationRequest)
```

to:

```c#
public async Task<ICommitableCredentialsResponse> PrepareResponseAsync(CredentialsRequest registrationRequest, CancellationToken cancellationToken)
```

So we can calculate things in parallel inside that method. I think the API interface would look much more natural.

Reply:
ACK, it is just called PrepareResponse it returns a Task

Reply:
We can remove `Commit` in the same go, it's no longer necessary now that `CredentialIssuer` is concurrency safe and more strict about duplicate serial numbers.